backend/src/services/hsm/hsm-service.ts

```diff
@@ -0,0 +1,349 @@
+
+      const decipher = s.createDecipher(algo, new graphene.Key(sessionKey).toType());
+      const outputBuffer = Buffer.alloc(ciphertext.length);
+
+      // Extract wrapped key, IV, and ciphertext
+      return decipher.once(ciphertext, outputBuffer);
+    };
+
+    if (providedSession) {
+      return $performDecryption(providedSession);
+    }
+    const decrypted = await withSession({ module, graphene }, (newSession) => $performDecryption(newSession));
+
+    return decrypted;
+  };
+
+  // We test the core functionality of the PKCS#11 module that we are using throughout Infisical. This is to ensure that the user doesn't configure a faulty or unsupported HSM device.
+  const $testPkcs11Module = async (session: grapheneLib.Session) => {
+    try {
+      if (!module) {
+        throw new Error("HSM service not initialized");
+      }
+
+      if (!session) {
+        throw new Error("Session not initialized");
+      }
+
+      const randomData = session.generateRandom(256);
+      const encryptedData = await encrypt(Buffer.from(randomData), session);
+      const decryptedData = await decrypt(encryptedData, session);
+
+      if (Buffer.from(randomData).toString("hex") !== Buffer.from(decryptedData).toString("hex")) {
+        throw new Error("Decrypted data does not match original data");
+      }
+
+      return true;
+    } catch (error) {
+      logger.error(error, "Error testing PKCS#11 module");
+      return false;
+    }
+  };
+
+  const isActive = async () => {
+    if (!module || !appCfg.isHsmConfigured) {
+      return false;
+    }
+
+    let pkcs11TestPassed = false;
+
+    try {
+      pkcs11TestPassed = await withSession({ module, graphene }, $testPkcs11Module);
+    } catch (err) {
+      logger.error(err, "isActive: Error testing PKCS#11 module");
+    }
+
+    return appCfg.isHsmConfigured && module !== null && pkcs11TestPassed;
+  };
+
+  const startService = async () => {
+    if (!appCfg.isHsmConfigured || !module) return;
+
+    try {
+      await withSession({ module, graphene }, async (session) => {
+        // Check if master key exists, create if not
+        if (!$keyExists(session)) {
+          // Generate 256-bit AES master key with persistent storage
+          session.generateKey(graphene.KeyGenMechanism.AES, {
+            class: graphene.ObjectClass.SECRET_KEY,
+            token: true,
+            valueLen: 256 / 8,
+            keyType: graphene.KeyType.AES,
+            label: appCfg.HSM_KEY_LABEL,
+            derive: true, // Enable key derivation
+            extractable: false,
+            sensitive: true,
+            private: true
+          });
+          logger.info(`Master key created successfully with label: ${appCfg.HSM_KEY_LABEL}`);
+        }
+
+        // Verify HSM supports required mechanisms
+        const mechs = session.slot.getMechanisms();
+        const mechNames: string[] = [];
+
+        // eslint-disable-next-line no-plusplus
+        for (let i = 0; i < mechs.length; i++) {
```

Comment:
Nit: Could've used `.map()` on `mechNames` would be cleaner.

Reply:
No. `MechanismCollection` doesn't implement an iterator.